File: server/fastapi/goban_server_fastapi/records/rest/create.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from fastapi import Depends
from goban_server_fastapi.auth import User, jwt_user
from goban_server_fastapi.db import DbClient, dictify
from goban_server_fastapi.records.models import Record
from goban_server_fastapi.rest import app


class CreateRequest(BaseModel):
    board_size: Literal[9, 13, 19]
    name: str = ""
    black_player: str = "Black"
    white_player: str = "White"
    comment: str = ""
    handicap: int = 0
    komi: float = 7.5
    ruleset: Literal["AGA", "JPN", "CHN"] = "AGA"


class CreateResponse(BaseModel):
    id: int
    owner: int
    board_size: Literal[9, 13, 19]
    created: datetime
    name: str
    black_player: str
    white_player: str
    comment: str
    handicap: int
    komi: float
    ruleset: Literal["AGA", "JPN", "CHN"]
    winner: Literal["U", "B", "W"]


@app.post(
    "/api/records/",
    status_code=201,
    response_model=CreateResponse,
)
def create_record(
    record_model: CreateRequest,
    db: DbClient = Depends(),
    current_user: User = Depends(jwt_user),
):
    record = Record(
        owner_id=current_user.id,
        created=datetime.now(),
        winner="U",
        **record_model.dict(),
    )
    with db.session() as session:
        session.add(record)
        session.commit()
        return {
            **dictify(record),
            "owner": record.owner_id,
        }

File: server/deno/src/record/create.ts
import { register, validator } from "/router.ts";
import { sql } from "/db.ts";
import { json_schema as J } from "/deps.ts";
import { getUserId } from "/auth/util.ts";

const CreateRequest = J.union(
  J.struct({
    board_size: J.number(),
  }),
)(
  J.partial({
    name: J.string(),
    black_player: J.string(),
    white_player: J.string(),
    comment: J.string(),
    handicap: J.number(),
    komi: J.number(),
    ruleset: J.string(),
  }),
);
const body = validator(CreateRequest);
const defaults = {
  name: "",
  black_player: "Black",
  white_player: "White",
  comment: "",
  handicap: 0,
  komi: 7.5,
  ruleset: "AGA",
};

register("POST", "/api/records/", async (request) => {
  const userId = await getUserId(request);
  const {
    board_size,
    name,
    black_player,
    white_player,
    comment,
    handicap,
    komi,
    ruleset,
  } = { ...defaults, ...await body(request) };
  if (![9, 13, 19].includes(board_size)) {
    return new Response("Invalid board size", { status: 400 });
  }
  const winner = "U";
  const now = new Date();
  const rows = await sql`INSERT INTO record_record
  (owner_id, board_size, name, black_player, white_player, comment, handicap, komi, ruleset, winner, created)
  VALUES
  (${userId}, ${board_size}, ${name}, ${black_player}, ${white_player}, ${comment}, ${handicap}, ${komi}, ${ruleset}, ${winner}, ${now})
  RETURNING id`;
  const recordId = Number.parseInt(rows[0].id, 10);
  return new Response(
    JSON.stringify({
      id: recordId,
      owner: userId,
      board_size,
      created: now,
      name,
      black_player,
      white_player,
      comment,
      handicap,
      komi,
      ruleset,
      winner,
    }),
    { status: 201 },
  );
});
